File: project1/project1.py
import networkx as nx
import sys
import time
import pandas as pd
import numpy as np
from scipy.special import loggamma
import matplotlib.pyplot as plt

MAX_PARENTS = 3

def get_parents(G, node):
    return [neighbor[0] for neighbor in G.in_edges(node)]

def get_children(G, node):
    return [neighbor[1] for neighbor in G.out_edges(node)]

def get_q(G, variables):
    return [np.prod([G.nodes[parent]['r'] 
        for parent in get_parents(G, var)], dtype=int) 
        for var in variables]

def get_r(G, variables):
    return [G.nodes[var]['r'] for var in variables]

def get_j_index(qi, parents_r_values, observed_parents_values, parents):
    one = np.array([1], dtype=int)
    cump = np.cumprod(parents_r_values[:-1])
    k = np.concatenate([one, cump])
    observed_parents_values = np.array(observed_parents_values) - 1
    return (np.dot(k, observed_parents_values)).astype(int)

def statistics(variables, G, D):
    n = D.shape[1]
    q = get_q(G, variables) 
    M = [np.zeros((q[i], G.nodes[variables[i]]['r']), dtype=int) 
        for i in range(n)]
    # print(q, [m.shape for m in M])
    for obs_i, obs in D.iterrows():
        for i,var in enumerate(variables):
            # use the observed value of variable as index into M 
            k = obs[var]-1 
            parents = get_parents(G, variables[i])
            j = 0
            if len(parents) > 0:
                j = get_j_index(q[i],
                    [G.nodes[parent]['r'] for parent in parents], 
                    [obs[parent] for parent in parents], parents)
            M[i][j,k] += 1
    return M

def prior(variables, G):
    n = len(variables)
    r = get_r(G, variables)
    q = get_q(G, variables)
    return [np.ones((q[i], r[i])) for i in range(n)]

def bayesian_score_component(M, alpha):
    p =  np.sum(loggamma(M + alpha)) 
    p -= np.sum(loggamma(alpha)) 
    p += np.sum(loggamma(np.sum(alpha, axis=1)))
    p -= np.sum(loggamma(np.sum(alpha, axis=1) + np.sum(M, axis=1)))
    return p

def bayesian_score(variables, G, D):
    n = len(variables)
    M = statistics(variables, G, D)
    alpha = prior(variables, G)
    return sum(bayesian_score_component(M[i], alpha[i]) for i in range(n))

def fit(variables, data, dataset_name, tic):
    G = nx.DiGraph()
    for i, v in enumerate(variables):
        G.add_node(v, r=max(data[v]), index=i)

    sorted_variables = list(nx.topological_sort(G))
    for k,i in enumerate(sorted_variables[1:]):
        y = bayesian_score(variables, G, data)
        while True:
            y_best, j_best = -np.inf, None
            for j in sorted_variables[:k]:
                if G.has_edge(j, i):
                    continue
                G.add_edge(j, i)
                y_new = bayesian_score(variables, G, data)
                if y_new > y_best:
                    y_best, j_best = y_new, j
                G.remove_edge(j, i)
            if y_best > y:
                print(f'best bayesian score: {y_best} (old: {y})')
                toc = time.perf_counter()
                print('Elapsed time:', toc-tic, 'seconds')
                if G.in_degree(i) > MAX_PARENTS:
                    # limit number of parents
                    break
                G.add_edge(j_best, i)
                y = y_best
                if dataset_name == 'large':
                    write_gph(G, {v:v for v in (variables)},
                        f"data/{dataset_name}.gph")
            else:
                break
    return G

def plot_graph(dataset_name, G=None):
    if G is None:
        edgelist = np.genfromtxt(f"data/{dataset_name}.gph", 
            delimiter=",", dtype=str)
        G = nx.DiGraph()
        G.add_edges_from(edgelist)

    for layer, nodes in enumerate(nx.topological_generations(G)):
        # `multipartite_layout` expects the layer as a node attribute, 
        # so add the numeric layer value as a node attribute
        for node in nodes:
            G.nodes[node]["layer"] = layer

    # Compute the multipartite_layout using the "layer" node attribute
    pos = nx.multipartite_layout(G, subset_key="layer")

    fig, ax = plt.subplots()
    nx.draw_networkx(G, pos=pos, ax=ax)
    ax.set_title(f"DAG layout in topological order for {dataset_name} dataset")
    fig.tight_layout()
    plt.savefig(f"data/{dataset_name}.png")
    plt.cla()
    plt.clf()

def write_gph(dag, idx2names, filename):
    with open(filename, 'w') as f:
        for edge in dag.edges():
            f.write("{}, {}\n".format(
                idx2names[edge[0]], idx2names[edge[1]]))

def compute(dataset_name):
    t0 = time.perf_counter()
    data = pd.read_csv(f"data/{dataset_name}.csv")
    variables = list(data.columns)
    G = fit(variables, data, dataset_name, t0)
    print(f'Done training {dataset_name} dataset! Final bayesian score:', 
        bayesian_score(variables, G, data))
    print('Elapsed time:', time.perf_counter() - t0, 'seconds')


def main():
    if len(sys.argv) != 2:
        raise Exception("usage: python project1.py <dataset_name>")
    compute(sys.argv[1])

    # if len(sys.argv) != 3:
    #     raise Exception("usage: python project1.py <infile>.csv <outfile>.gph")
    # inputfilename = sys.argv[1]
    # outputfilename = sys.argv[2]
    # compute(inputfilename, outputfilename)

if __name__ == '__main__':
    main()

File: project1/project1.jl
using Graphs
using Printf
using CSV
using DataFrames

"""
    write_gph(dag::DiGraph, idx2names, filename)

Takes a DiGraph, a Dict of index to names and a output filename to write the graph in `gph` format.
"""
function write_gph(dag::DiGraph, idx2names, filename)
    open(filename, "w") do io
        for edge in edges(dag)
            @printf(io, "%s,%s\n", idx2names[src(edge)], idx2names[dst(edge)])
        end
    end
end


function compute(infile, outfile)
    df = CSV.read("example.csv", DataFrame)

    # WRITE YOUR CODE HERE
    # FEEL FREE TO CHANGE ANYTHING ANYWHERE IN THE CODE
    # THIS INCLUDES CHANGING THE FUNCTION NAMES, MAKING THE CODE MODULAR, BASICALLY ANYTHING

end

#################### representation 1
struct Variable
	name::Symbol
	r::Int # number of possible values
end

const Assignment = Dict{Symbol,Int}
const FactorTable = Dict{Assignment,Float64}

struct Factor
	vars::Vector{Variable}
	table::FactorTable
end

variablenames(ϕ::Factor) = [var.name for var in ϕ.vars]

select(a::Assignment, varnames::Vector{Symbol}) =
	Assignment(n=>a[n] for n in varnames)

function assignments(vars::AbstractVector{Variable})
    names = [var.name for var in vars]
    return vec([Assignment(n=>v for (n,v) in zip(names, values))
    			for values in product((1:v.r for v in vars)...)])
end

function normalize!(ϕ::Factor)
	z = sum(p for (a,p) in ϕ.table)
	for (a,p) in ϕ.table
		ϕ.table[a] = p/z
	end
	return ϕ
end
####################

#################### parameter-learning 1
# The sub2ind(siz, x) function returns a linear index into an array 
# with dimensions specified by siz given coordinates x. 
# It is used to identify which parental instantiation is 
# relevant to a particular data point and variable.

function sub2ind(siz, x)
    k = vcat(1, cumprod(siz[1:end-1]))
    return dot(k, x .- 1) + 1
end

function statistics(vars, G, D::Matrix{Int})
    n = size(D, 1)
    r = [vars[i].r for i in 1:n]
    q = [prod([r[j] for j in inneighbors(G,i)]) for i in 1:n]
    M = [zeros(q[i], r[i]) for i in 1:n]
    for o in eachcol(D)
        for i in 1:n
            k = o[i]
            parents = inneighbors(G,i)
            j = 1
            if !isempty(parents)
                 j = sub2ind(r[parents], o[parents])
            end
            M[i][j,k] += 1.0
        end
    end
    return M
end
####################

#################### parameter-learning 2
function prior(vars, G)
    n = length(vars)
    r = [vars[i].r for i in 1:n]
    q = [prod([r[j] for j in inneighbors(G,i)]) for i in 1:n]
    return [ones(q[i], r[i]) for i in 1:n]
end
####################

#################### structure-learning 1
function bayesian_score_component(M, α)
    p =  sum(loggamma.(α + M))
    p -= sum(loggamma.(α))
    p += sum(loggamma.(sum(α,dims=2)))
    p -= sum(loggamma.(sum(α,dims=2) + sum(M,dims=2)))
    return p
end

function bayesian_score(vars, G, D)
    n = length(vars)
    M = statistics(vars, G, D)
    α = prior(vars, G)
    return sum(bayesian_score_component(M[i], α[i]) for i in 1:n)
end
####################

#################### inference 6
function topological_sort(G)
	G = deepcopy(G)
	ordering = []
	parentless = filter(i -> isempty(inneighbors(G, i)), 1:nv(G))
	while !isempty(parentless)
		i = pop!(parentless)
		push!(ordering, i)
		for j in copy(outneighbors(G, i))
			rem_edge!(G, i, j)
			if isempty(inneighbors(G, j))
				push!(parentless, j)
			end
		end
	end
	return ordering
end
####################

#################### structure-learning 2
struct K2Search
    ordering::Vector{Int} # variable ordering
end

function fit(method::K2Search, vars, D)
    G = SimpleDiGraph(length(vars))
    for (k,i) in enumerate(method.ordering[2:end])
        y = bayesian_score(vars, G, D)
        while true
            y_best, j_best = -Inf, 0
            for j in method.ordering[1:k]
                if !has_edge(G, j, i)
                    add_edge!(G, j, i)
                    y′ = bayesian_score(vars, G, D)
                    if y′ > y_best
                        y_best, j_best = y′, j
                    end
                    rem_edge!(G, j, i)
                end
            end
            if y_best > y
                y = y_best
                add_edge!(G, j_best, i)
            else
                break
            end
        end
    end
    return G
end
####################


if length(ARGS) != 2
    error("usage: julia project1.jl <infile>.csv <outfile>.gph")
end

inputfilename = ARGS[1]
outputfilename = ARGS[2]

compute(inputfilename, outputfilename)
